### PR TITLE
Fixes #1520 by using explicit left join when sorting over relationships

### LINKED ISSUE #1520: Hibernate's implicit inner joins mess up pagination
## Expected Behavior
When sorting over nullable relationship, there is correct match between number of records and pagination

## Current Behavior
When sorting over nullable relationship, actual query will use inner/cross join while count query won't.
This results in wrong pagination, where last pages based on count does not contain any data

## Possible Solution
Based on https://lists.jboss.org/pipermail/hibernate-dev/2018-February/017299.html there is no will in hibernate to fix this.
Suggested solution is to make explicit joins instead

## Steps to Reproduce (for bugs)
http://api.faforever.com/data/mod?sort=latestVersion.reviewsSummary.lowerBound&page%5Bsize%5D=20&page%5Btotals%5D&page%5Bnumber%5D=20
There is 374 records when inner joining but 1300 without. This results in empty responses from page 20

## Your Environment

* Elide version used: 4.6.6
* Environment name and version (Java 1.8.0_152):
* Operating System and version:
* Link to your project: https://github.com/FAForever/faf-java-api/issues/415

### PR DESCRIPTION
**Pull request by bukajsytlos**
Resolves #1520

## Description
When resolving hibernate query add explicit joins based on sort paths

## Motivation and Context
See issue

## How Has This Been Tested?
Just by debugging generated query.
I would like to do some unit/IT tests, but didn't figured how can I debug elide while running IT.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by aklish** [comment]
I suspect this will also fixes #920.  

Thanks for the submission.  Will review later today.

**Comment by aklish** [comment]
In IntelliJ, you can debug IT tests with a particular data store by configuring the DataStoreHarness and setting the classpath like this: 

Screen Shot 2020-09-04 at 10 59 22 AM

File: elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/hibernate/hql/RootCollectionFetchQueryBuilder.java [hunk 1]
```diff
@@ -49,7 +49,8 @@ public Query build() {
 
             //Build the JOIN clause
             String joinClause =  getJoinClauseFromFilters(filterExpression.get())
-                    + extractToOneMergeJoins(entityClass, entityAlias);
+                    + extractToOneMergeJoins(entityClass, entityAlias)
+                    + explicitSortJoins(sorting, entityClass);
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/hibernate/hql/RootCollectionFetchQueryBuilder.java:53]: You will need to make similar changes to SubCollectionFetchQueryBuilder.java

Also, I recommend adding a test on a subcollection: /foo/123/bar where your new sorting join logic can be validated (sort on bar.blah.someField or something).

**Comment by bukajsytlos** [comment]
@aklish thanks. I am now kind of confused. I have just now realized that there is already merged PR for #920. So how should I proceed?

**Comment by aklish** [comment]
Apologies.  I had forgotten we had fixed this in the elide-5.x branch already.  I would recommend we back port that change in elide 4 (master).

**Comment by bukajsytlos** [comment]
Should I back port it?

**Comment by aklish** [comment]
I was about to push some changes to your branch.  I kept your code, but back ported the tests.  There were a few small bug fixes I found when I back ported the tests.  Check back in 15 minutes.

**Comment by aklish** [comment]
Looks like I can't push to your fork.  

To github.com:bukajsytlos/elide.git
 ! [remote rejected]     elide-1520-wrong-pagination-sort-over-nullable-relationship -> elide-1520-wrong-pagination-sort-over-nullable-relationship (permission denied)
error: failed to push some refs to 'git@github.com:bukajsytlos/elide.git'

**Comment by aklish** [comment]
Here is a patch file for the same:

[patch.txt](https://github.com/yahoo/elide/files/5211318/patch.txt)

**Comment by bukajsytlos** [comment]
thanks. I have incorporated your patch with minor tweak

**Review by aklish — APPROVED** [review]

**Comment by bukajsytlos** [comment]
Hi @aklish. Please any schedule on releasing newer version?

**Comment by aklish** [comment]
Likely today.

**Comment by aklish** [comment]
Released.

**Comment by bukajsytlos** [comment]
thank you ;)

**Pull request merged**